### ISSUE TITLE
Remove bitcoin dependency

### ISSUE THREAD

**Issue by palango** [raiden-network/raiden#1200]
## Problem Definition

We use the `bitcoin` library for converting a private to a public key in [`raiden/utils/crypto.py`](https://github.com/raiden-network/raiden/blob/4493a19041c73959267518f055d078f3357f4528/raiden/utils/crypto.py#L3). The `bitcoin` dependency is not longer maintained and should be removed.

## Solution

Find a replacement for the `bitcoin.privtopub` function and remove the `bitcoin` dependency.

## Tasklist

- [ ] Find replacement for the `bitcoin.privtopub` function
- [ ] Remove the `bitcoin` dependency

**Comment by aneeshvaidya** [comment]
Hi! I'm a first time contributor, would it be okay if took up this task?

**Comment by LefterisJP** [comment]
@aneeshvaidya Sure go ahead!

**Comment by aneeshvaidya** [comment]
Cool! I'll research a bit and update here with a potential replacement

**Comment by andrevmatos** [comment]
[µRaiden uses an implementation which requires only coincurve](https://github.com/raiden-network/microraiden/blob/fd81a612e671956166f114f34f9aa6f7ab209471/microraiden/microraiden/utils/crypto.py#L32-L35), it may be worth checking.

**Comment by aneeshvaidya** [comment]
@andrevmatos coincurve seems to have what's needed. I'm also curious whether there's an issue with the Raiden team just forking the `bitcoin` library, and having the community perform maintenance/updates there.

**Comment by palango** [comment]
@aneeshvaidya I guess it doesn't make sense for the raiden team to maintain the bitcoin library when we're only using one function.

Following the implementation from microraiden should be fine, you only need to find some replacements for helper functions like `remove_0x_prefix`.

**Comment by aneeshvaidya** [comment]
It seems like using the implementation from microraiden would work, but it makes use of the helpers from `eth_utils` quite a bit. Another alternative I've found is `py_ecc` which is maintained by Ethereum, and has the same implementation as the `bitcoin` package.

**Comment by palango** [comment]
Instead of adding another dependency I think we should go for an implementation similar to the one in microraiden. We have some helper functions available in the [`utils` module](https://github.com/raiden-network/raiden/blob/master/raiden/utils/__init__.py). Let us know when some helper function is missing.

**Comment by ofek** [comment]
I can help if need be!

**Comment by aneeshvaidya** [comment]
@ofek I was able to replicate `bitcoin.privtopub` by calling like this: 

`PrivateKey.from_hex(private).public_key.format(compressed=False)`

 However, I am seeing that the current implementation in raiden calls `bitcoin.encode_pubkey` with the `bin_electrum` format. Based on the output difference, the coincurve formatting has 1 byte appended to the beginning. Should I just return the last 64 bytes? Also, is the private key guaranteed to be a hex string as it appears to be in the microraiden implementation linked above?

**Comment by ofek** [comment]
Yes you can, the last 64 bytes are the 2 points. I'm not sure how Raiden passes around private keys, but if that works, then it must be hex :slightly_smiling_face:

Wow, pybitcointools is horrible to read

https://github.com/vbuterin/pybitcointools/blob/aeb0a2bbb8bbfe421432d776c649650eaeb882a5/bitcoin/main.py#L291-L301

https://github.com/vbuterin/pybitcointools/blob/aeb0a2bbb8bbfe421432d776c649650eaeb882a5/bitcoin/py3specials.py#L80-L99

https://github.com/vbuterin/pybitcointools/blob/aeb0a2bbb8bbfe421432d776c649650eaeb882a5/bitcoin/main.py#L187

**Comment by aneeshvaidya** [comment]
If the encoding is guaranteed to be hex, then we don't even need to replicate the `encode` function in pybitcointools. Need someone more familiar with the codebase to confirm though.

**Comment by LefterisJP** [comment]
@aneeshvaidya You opened two PRs and closed them again. Why? You can edit PRs by fore pushing on the PR's branch.

**Comment by aneeshvaidya** [comment]
Sorry for that, there were some issues and I didn't know you could do that. For some reason a force pushed commit on a branch doesn't let me reopen a PR. Apologies for the noise!

**Comment by palango** [comment]
@aneeshvaidya  removed all direct calls to the `bitcoin`library in #1228. However `ethereum.keys` still imports `bitcoin` and shows a warning if it isn't available. Therefore the dependency wasn't removed yet.

**Comment by aneeshvaidya** [comment]
@palango I can explore replacing that utility if it makes sense to

**Comment by palango** [comment]
@aneeshvaidya The problem is here https://github.com/ethereum/pyethereum/blob/develop/ethereum/tools/keys.py#L18-L25

We can remove all usages to `ethereum.keys` but it still might be used internally. And other implementations like https://github.com/ethereum/eth-account don't seem to be ready for primetime yet.

**Comment by LefterisJP** [comment]
@palango: So do I understand correctly that this issue is going to be addressed when we switch to web3.py?

**Comment by palango** [comment]
Not necessarily. There is some code to check key files for which there is no replacement atm. But that should be easy to remove or to reimplement ourselves.

**Comment by pcppcp** [comment]
Hm, if you need to read keystore maybe check https://github.com/raiden-network/microraiden/blob/master/microraiden/utils/private_key.py

IIRC it doesn't require `bitcoin` (and `ethereum` seems to use `bitcoin` only in one specific function that I don't see used anywhere else)

**Comment by palango** [comment]
Yes, these are already used. However `ethereum.keys` still imports bitcoin and shows a warning if it isn't available.

**Comment by LefterisJP** [comment]
We really should just drop `ethereum.tools.keys` then.

The 2 functions we use from there are:

```python
def check_keystore_json(jsondata):
    """Check if ``jsondata`` has the structure of a keystore file version 3.

    Note that this test is not complete, e.g. it doesn't check key derivation or cipher parameters.

    :param jsondata: dictionary containing the data from the json file
    :returns: `True` if the data appears to be valid, otherwise `False`
    """
    if 'crypto' not in jsondata and 'Crypto' not in jsondata:
        return False
    if 'version' not in jsondata:
        return False
    if jsondata['version'] != 3:
        return False
    crypto = jsondata.get('crypto', jsondata.get('Crypto'))
    if 'cipher' not in crypto:
        return False
    if 'ciphertext' not in crypto:
        return False
    if 'kdf' not in crypto:
        return False
    if 'mac' not in crypto:
        return False
    return True
```
and

```python
def decode_keystore_json(jsondata, pw):
    # Get KDF function and parameters
    if "crypto" in jsondata:
        cryptdata = jsondata["crypto"]
    elif "Crypto" in jsondata:
        cryptdata = jsondata["Crypto"]
    else:
        raise Exception("JSON data must contain \"crypto\" object")
    kdfparams = cryptdata["kdfparams"]
    kdf = cryptdata["kdf"]
    if cryptdata["kdf"] not in kdfs:
        raise Exception("Hash algo %s not supported" % kdf)
    kdfeval = kdfs[kdf]["calc"]
    # Get cipher and parameters
    cipherparams = cryptdata["cipherparams"]
    cipher = cryptdata["cipher"]
    if cryptdata["cipher"] not in ciphers:
        raise Exception("Encryption algo %s not supported" % cipher)
    decrypt = ciphers[cipher]["decrypt"]
    # Compute the derived key
    derivedkey = kdfeval(pw, kdfparams)
    assert len(derivedkey) >= 32, \
        "Derived key must be at least 32 bytes long"
    # print(b'derivedkey: ' + encode_hex(derivedkey))
    enckey = derivedkey[:16]
    # print(b'enckey: ' + encode_hex(enckey))
    ctext = decode_hex(cryptdata["ciphertext"])
    # Decrypt the ciphertext
    o = decrypt(ctext, enckey, cipherparams)
    # Compare the provided MAC with a locally computed MAC
    # print(b'macdata: ' + encode_hex(derivedkey[16:32] + ctext))
    mac1 = sha3(derivedkey[16:32] + ctext)
    mac2 = decode_hex(cryptdata["mac"])
    if mac1 != mac2:
        raise ValueError("MAC mismatch. Password incorrect?")
    return o
```

The first function `check_keystore_json()` is ridiculously simple and we can just copy it as is.
The second function, `decode_keystore_json()` loads and decodes the keys. We can use [`eth-keys`](https://github.com/ethereum/eth-keys) for that.

**Comment by palango** [comment]
@aneeshvaidya The solution for the last step of removing `bitcoin` are specified above. Do you still wanna work on this?

**Comment by aneeshvaidya** [comment]
@palango Yes! I'll take a look through this and submit another PR

**Comment by palango** [comment]
@aneeshvaidya Do you have time to fix this, otherwise I'll fix it.

**Comment by aneeshvaidya** [comment]
I don't believe I'll get around to this for a while, so go ahead

**Comment by andrevmatos** [comment]
Once we finish #1293, we may look at using [web3.py methods for opening private keys](http://web3py.readthedocs.io/en/stable/web3.eth.account.html#extract-private-key-from-geth-keyfile) directly.
